File: helpers/sample_sorting_methods/quick_sort.py

def quick_sort(input_list, low, high):
	"""
	
	:param input_list: input list that needs to be sorted
	:return: sorted list
	"""
	if low < high:
		# take low and high indices of the input_list
		# partition the input_list and sort, assuming pivot element is highest element
		pivot_idex = partition(input_list=input_list, low_index=low, high_index=high)
		# get the index of pivot element run quick_sort on elements smaller than pivot element
		print('sorting items larger than pivot element. pivot index: {}'.format(pivot_idex))
		quick_sort(input_list=input_list, low=low, high=pivot_idex-1)
		# get the index of pivot element, run quick_sort on elements bigger than pivot element
		print('sorting items larger than pivot element. pivot index: {}'.format(pivot_idex))
		quick_sort(input_list=input_list, low=pivot_idex+1, high=high)
	
	return input_list
	
def partition(input_list, low_index, high_index):
	"""
	
	:param input_list:
	:param low_index:
	:param high_index:
	:return:
	"""
	# take pivot index as the last item
	pivot_index = low_index - 1
	# save pivot element
	pivot_element = input_list[pivot_index]
	
	# since nth element is the pivot element of input list,
	# we must iterate over input list from 0 to n-1
	for index in range(low_index, high_index-1):
		# compare input list elements with pivot element
		# if pivot element is bigger than input_list's element, swap them.
		if input_list[index] <= pivot_element:
			pivot_index += 1
			print('swapping: {} with {}'.format(input_list[index], input_list[pivot_index+1]))
			input_list[index], input_list[pivot_index] = input_list[pivot_index], input_list[index]
	print('pivot_index after: {}'.format(pivot_index))
	input_list[pivot_index+1], input_list[high_index] = input_list[high_index], input_list[pivot_index+1]
	#print 'input list: {}\n'.format(input_list)
	return pivot_index



#input_list = [[1, 5, 3, 5, 6, 7, 99], [5, 1, 4, 3, 8, 99], [1, 1, 1, 1, 1, 1], [5, -1, 4, 3, 8, 99], [12, 11, 13, 5, 6],
#              None, [], [1, 2, 3, 4, 5], [10, 7, 8, 9, 1, 5]]

input_list = [[10, 80, 30, 90, 40, 50, 70]]


def sorting_ut():
	for test_list in input_list:
		print('**********************************')
	# 	print('Input list before sorting:       {}'.format(test_list))
	# 	print('Input list after selection sort: {}'.format(sorting_methods.selection_sort(input_list=test_list)))
	# 	print('Input list after bubble sort:    {}'.format(sorting_methods.bubble_sort(input_list=test_list)))
	# 	print('Input list after insertion sort: {}\n'.format(sorting_methods.insertion_sort(input_list=test_list)))
	#	print('Input list after merge sort:     {}'.format(sorting_methods.merge_sort(input_list=test_list)))
	# input_list = [10, 7, 8, 9, 1, 5]
		print('Merge sort: {}'.format(quick_sort(input_list=test_list, low=0, high=len(test_list)-1)))
		print('**********************************\n\n')


sorting_ut()

File: C/sortingMethods/quickSort.c

#include "sortingMethods.h"

void quick_sort(int input_arr[], int arr_size, int low, int high);
int partition(int input_arr[], int arr_size, int low, int high);

/* Function to print an array */
void printArray(int arr[], int size) 
{ 
    int i;
    printf("[");
    for (i=0; i < size; i++) 
        printf("%d ", arr[i]); 
    printf("]\n");
} 

// A utility function to swap two elements 
void swap(int* a, int* b) 
{ 
    int t = *a; 
    *a = *b; 
    *b = t; 
}

/* This function takes last element as pivot, places 
   the pivot element at its correct position in sorted 
   array, and places all smaller (smaller than pivot) 
   to left of pivot and all greater elements to right 
   of pivot */
int partition(int input_arr[], int arr_size, int low, int high){

    int i = low;
    int pivot = input_arr[high];

    for(int j = low; j < high; j++){
        if(pivot >= input_arr[j]){
            swap(&input_arr[j],&input_arr[i]);
            i++;
        }
    }

    printArray(input_arr, )

    swap(&input_arr[i], &input_arr[high]);

    printf("2. [");
    for(int i=0; i<5; i++){
        printf("%d,", input_arr[i]);
    }
    printf("]\n");

    return i;

}

void quick_sort(int input_arr[], int arr_size, int low, int high){

    if(low<high){
        int pivot_index = partition(input_arr, arr_size, low, high);
        quick_sort(input_arr, arr_size, ow, pivot_index-1);
        quick_sort(input_arr, arr_size, pivot_index+1, high);
    }

}

int main(void){

    int arr_size = 5;
    int input_arr[5] = {10,40,80,70,50};

    printf("Before: [");
    for(int i=0; i<5; i++){
        printf("%d,", input_arr[i]);
    }
    printf("]\n");

    quick_sort(input_arr, arr_size, 0, 4);

    printf("After: [");
    for(int i=0; i<5; i++){
        printf("%d,", input_arr[i]);
    }
    printf("]\n");

    return 0;
}
